server/tserver/src/main/java/org/apache/accumulo/tserver/NativeMap.java

```diff
@@ -304,7 +304,13 @@ private class NMIterator implements Iterator<Map.Entry<Key,Value>> {
       hasNext = nmiPointer != 0;
 
       nmiPtr.set(nmiPointer);
-      cleanableNMI = NativeMapCleanerUtil.deleteNMIterator(this, nmiPtr);
+      if (nmiPointer == 0) {
+        // registering a cleaner takes resources and may lock, so only bother registering if there
+        // is something to delete
+        cleanableNMI = null;
+      } else {
+        cleanableNMI = NativeMapCleanerUtil.deleteNMIterator(this, nmiPtr);
+      }
```

Comment:
Implementation looks fine, but I think it can be written more concisely as:

```suggestion
      // avoid registering a cleanable if there's nothing to delete
      cleanableNMI = hasNext ? NativeMapCleanerUtil.deleteNMIterator(this, nmiPtr) : null;
```